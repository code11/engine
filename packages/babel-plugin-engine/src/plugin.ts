import * as Babel from "@babel/core";
import { VariableDeclaratorVisitor } from "./visitors";

export type PluginConfig = {
  viewLibrary: string;
};

type pluginType = (babel: typeof Babel, state: PluginConfig) => {};

export const plugin: pluginType = (babel, state) => {
  return {
    name: "babel-plugin-engine",
    visitor: {
      VariableDeclarator: VariableDeclaratorVisitor(babel, state),
    },
  };
};
