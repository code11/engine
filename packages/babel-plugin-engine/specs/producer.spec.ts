import pluginTester from "babel-plugin-tester";
import prettier from "prettier";
import plugin from "../src";

const config = require("./babelConfig.json");

// Add tests regarding let foo: producer = () => {}, bar: producer = () => {}

// Test invalid syntax with const foo: view = 123

pluginTester({
  plugin,
  babelOptions: { filename: __filename, ...config },
  formatResult: (result: any) => {
    return prettier.format(result, {
      parser: "babel",
    });
  },
  pluginName: "babel-plugin-engine",
  tests: {
    "should compile a producer": {
      code: `
      const foo: producer = ({ foo = observe.foo }) => {}
      `,
      snapshot: true,
    },
    "should throw if not given an arrow function": {
      code: `const foo: producer = 123`,
      error: true,
    },
    "should throw if not function param is not an object": {
      code: `const foo: producer = (foo) => {}`,
      error: true,
    },
    "should throw an error if it is not invoked with an arrow function": {
      code: `
        const result: producer = {
          foo = '123'
        }
      `,
      error: true,
    },
  },
});
