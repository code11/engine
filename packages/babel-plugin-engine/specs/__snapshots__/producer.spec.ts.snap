// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`babel-plugin-engine should compile a producer: should compile a producer 1`] = `

const foo: producer = ({ foo = observe.foo }) => {}

      ↓ ↓ ↓ ↓ ↓ ↓

const foo = {
  props: {
    type: "STRUCT",
    value: {
      foo: {
        type: "OBSERVE",
        path: [
          {
            type: "CONST",
            value: "foo",
          },
        ],
      },
    },
    meta: {},
  },
  meta: {},
  fn: ({ foo: foo }) => {},
};


`;
